README.md
# H-99 - Haskell 99 Questions


CHANGELOG.md
# Changelog for `h99`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/Spec.hs
{-# LANGUAGE ScopedTypeVariables #-}
import Test.Hspec
import Test.QuickCheck

import System.IO

import Control.Exception (evaluate)
import Control.Monad (replicateM)

import Data.Function ((&))
import Data.List (genericLength, genericReplicate, genericTake, intercalate)
import Data.List.Split (splitOn)

import qualified Combinatorics

import Lib

main :: IO ()
main = hspec $ do
  -- Problem 1
  describe "myLast" $ do
    it "returns the last element of the list" $ do
      myLast [1, 2, 3, 4] `shouldBe` 4
      myLast ['x', 'y', 'z'] `shouldBe` 'z'
      evaluate (myLast []) `shouldThrow` errorCall "empty list"

  -- Problem 2
  describe "myButLast" $ do
    it "returns the last but one element of a list" $ do
      myButLast [1, 2, 3, 4] `shouldBe` 3
      myButLast ['a'..'z'] `shouldBe` 'y'
      evaluate (myButLast []) `shouldThrow` errorCall "empty list"
      evaluate (myButLast [1]) `shouldThrow` errorCall "list with one element"

  -- Problem 3
  describe "elementAt" $ do
    it "returns the K'th element of a list, 1-indexed" $ do
      elementAt [1,2,3] 2 `shouldBe` 2
      elementAt "haskell" 5 `shouldBe` 'e'

  -- Problem 4
  describe "myLength" $ do
    it "returns the number of elements of a list" $ do
      myLength [123, 456, 789] `shouldBe` 3
      myLength "Hello, world!" `shouldBe` 13
      myLength [] `shouldBe` 0

  -- Problem 5
  describe "myReverse" $ do
    it "returns the reversed copy of the list" $ do
      myReverse [] `shouldBe` ([] :: [Int])
      myReverse "A man, a plan, a canal, panama!" `shouldBe` "!amanap ,lanac a ,nalp a ,nam A"
      myReverse [1,2,3,4] `shouldBe` [4,3,2,1]

  -- Problem 6
  describe "isPalindrome" $ do
    it "returns whether a list is a palindrome" $ do
      isPalindrome [1,2,3] `shouldBe` False
      isPalindrome "madamimadam" `shouldBe` True
      isPalindrome [1,2,4,8,16,8,4,2,1] `shouldBe` True

  -- Problem 7
  describe "flatten" $ do
    it "flattens a nested list structure" $ do
      flatten (Elem 5) `shouldBe` [5]
      flatten (List [Elem 1, List [Elem 2, List [Elem 3, Elem 4], Elem 5]]) `shouldBe` [1,2,3,4,5]
      flatten (List []) `shouldBe` ([] :: [Int])

  -- Problem 8
  describe "compress" $ do
    it "eliminates consecutive duplicates of list elements." $ do
      compress "aaaabccaadeeee" `shouldBe` "abcade"
      compress [] `shouldBe` ([] :: [Int])

  -- Problem 9
  describe "pack" $ do
    it "packs consecutive duplicates of list elements into sublists" $ do
      pack "aaaabccaadeeee" `shouldBe` ["aaaa","b","cc","aa","d","eeee"]

  -- Problem 10
  describe "encode" $ do
    it "returns the run-length encoding of a list" $ do
      encode "aaaabccaadeeee" `shouldBe` [(4,'a'),(1,'b'),(2,'c'),(2,'a'),(1,'d'),(4,'e')]

  -- Problem 11
  describe "encodeModified" $ do
    it "packages the encoding into custom datatype" $ do
      encodeModified "aaaabccaadeeee" `shouldBe` [Multiple 4 'a',Single 'b',Multiple 2 'c', Multiple 2 'a',Single 'd',Multiple 4 'e']

  -- Problem 12
  describe "decodeModified" $ do
    it "decodes the custom datatype of the run-length encoding" $ do
      decodeModified [Multiple 4 'a',Single 'b',Multiple 2 'c',Multiple 2 'a',Single 'd',Multiple 4 'e'] `shouldBe` "aaaabccaadeeee"

  -- Problem 13
  describe "encodeDirect" $ do
    it "run-length encodes a list directly" $ do
      encodeDirect "aaaabccaadeeee" `shouldBe` [Multiple 4 'a',Single 'b',Multiple 2 'c',Multiple 2 'a',Single 'd',Multiple 4 'e']

  -- Problem 14
  describe "dupli" $ do
    it "duplicate the elements of a list" $ do
      dupli [1, 2, 3] `shouldBe` [1,1,2,2,3,3]

  -- Problem 15
  describe "repli" $ do
    it "replicates the elements of a list a given number of times" $ do
       repli "abc" 3 `shouldBe` "aaabbbccc"

  -- Problem 16
  describe "dropEvery" $ do
    it "drops every n-th element from a list" $ do
      dropEvery "abcdefghik" 3 `shouldBe` "abdeghk"
  
  -- Problem 17
  describe "split" $ do
    it "splits a list into two parts; the length of the first part is given." $ do
      split "abcdefghik" 3 `shouldBe` ("abc", "defghik")

  -- Problem 18
  describe "slice" $ do
    it "extracts a slice from a list" $ do
      slice "abcdefghik" 3 7 `shouldBe` "cdefg"

  -- Problem 19
  describe "rotate" $ do
    it "rotates a list n places to the left" $ do
      rotate "abcdefgh" 3 `shouldBe` "defghabc"
      rotate "abcdefgh" (-2) `shouldBe` "ghabcdef"

  -- Problem 20
  describe "removeAt" $ do
    it "removes the k-th element from a list" $ do
      removeAt 2 "abcd" `shouldBe` ('b',"acd")

  -- Problem 21
  describe "insertAt" $ do
    it "inserts an element at a given position into a list" $ do
      insertAt 'X' "abcd" 2 `shouldBe` "aXbcd"

  -- Problem 22
  describe "range" $ do
    it "creates a list containing integers within a given ragne" $ do
      range 4 9 `shouldBe` [4,5,6,7,8,9]

  -- Problem 23
  describe "rnd_select" $ do
    it "Extract a given number of randomly selected elements from a list" $ do
      let xs = "1234567"
      rnd_select xs 3 >>= (`shouldSatisfy` (\results -> length results == 3 && all (`elem` xs) results))
      

  -- Problem 24
  describe "diff_select" $ do
    it "draws n different random numbers from the set {1..m}" $ do
      diff_select 6 49 >>= (`shouldSatisfy` (\results -> length results == 6 && all (`elem` [1..49]) results))

  -- Problem 25
  describe "rnd_permu" $ do
    it "generates a random permutation of the elements of a list" $ do
      let source = "abcdef"
      rnd_permu source >>= (`shouldSatisfy` (\results -> length results == length source && all (`elem` source) results))

  -- Problem 26
  describe "combinations" $ do
    it "generates the combinations of k distinct objects from the n-element list" $ do
      let source = "abcdef"
      let results = combinations 3 source
      length results `shouldBe` Combinatorics.binomial (length "abcdef") 3
      results `shouldBe` Combinatorics.tuples 3 source

  -- Problem 27
  describe "group" $ do
    it "groups the elements of a set into disjoint subsets." $ do
      group [2,3,4] ["aldo","beat","carla","david","evi","flip","gary","hugo","ida"] `shouldSatisfy` (\results -> length results == 1260)
      group [2,2,5] ["aldo","beat","carla","david","evi","flip","gary","hugo","ida"] `shouldSatisfy` (\results -> length results == 756)

  -- Problem 28
  describe "Sorting a list of lists according to length of sublists" $ do
    describe "lsort" $ do
      it "sorts sublists according to their length" $ do
        lsort ["abc","de","fgh","de","ijkl","mn","o"] `shouldBe` ["o","de","de","mn","abc","fgh","ijkl"]
    
    describe "lfsort" $ do
      it "sorts sublists according to their length frequency" $ do
        lfsort ["abc", "de", "fgh", "de", "ijkl", "mn", "o"] `shouldBe` ["ijkl","o","abc","fgh","de","de","mn"]

  -- Problem 31
  describe "isPrime" $ do
    it "determines whether a given integer is prime" $ do
      isPrime 7 `shouldBe` True

  -- Problem 32
  describe "myGCD" $ do
    it "determines the greatest common divisor of two positive integer numbers" $ do
      myGCD 36 63 `shouldBe` 9
      myGCD (-3) (-6) `shouldBe` 3
      myGCD (-3) 6 `shouldBe` 3

  -- Problem 33
  describe "coprime" $ do
    it "determines whether two positive integer numbers are coprime" $ do
      coprime 35 64 `shouldBe` True

  -- Problem 34
  describe "totient" $ do
    it "calculates Euler's totient function phi(m)." $ do
      totient 10 `shouldBe` 4
  
  -- Problem 35
  describe "primeFactors" $ do
    it "determines the prime factors of a given positive integer" $ do
      primeFactors 315 `shouldBe` [3, 3, 5, 7]

  -- Problem 36
  describe "prime_factors_mult" $ do
    it "determines the prime factors of a given positive integer" $ do
      prime_factors_mult 315 `shouldBe` [(3,2),(5,1),(7,1)]

  -- Problem 37
  describe "totient'" $ do
    it "calculates Euler's totient function phi(m) (improved)." $ do
      totient' 10 `shouldBe` 4

  -- Problem 39
  describe "primesR" $ do
    it "lists out the prime numbers between lo and hi" $ do
      primesR 10 20 `shouldBe` [11,13,17,19]

  -- Problem 40
  describe "goldbach" $ do
    it "finds the pair of primes that sums up to a even number n" $ do
      goldbach 28 `shouldBe` (5, 23)

  -- Problem 41
  describe "Goldbach's list" $ do
    describe "goldbachList" $ do
      it "finds the pairs of primes that sum up to a even number n in range lo and hi" $ do
        goldbachList 9 20 `shouldBe` [(3,7),(5,7),(3,11),(3,13),(5,13),(3,17)]

    describe "goldbachList'" $ do
      it "finds the pairs of primes that sum up to a even number n in range lo and hi" $ do
        goldbachList' 4 2000 50 `shouldBe` [(73,919),(61,1321),(67,1789),(61,1867)]

  describe "Logic" $ do
    -- Problem 46
    describe "and'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `and'` b) == (a && b)

    describe "or'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `or'` b) == (a || b)

    describe "nand'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `nand'` b) == (not $ a && b)

    describe "nor'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `nor'` b) == (not $ a || b)

    describe "xor'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `xor'` b) == (a /= b)

    describe "impl'" $ do
      it "has correct truth table" $ do
        True `impl'` True `shouldBe` True
        True `impl'` False `shouldBe` False
        False `impl'` True `shouldBe` True
        False `impl'` False `shouldBe` True
  
    describe "equ'" $ do
      it "has correct truth table" $ do
        property $ \(a :: Bool) (b :: Bool) -> (a `equ'` b) == (a == b)

    let parse :: String -> [[Bool]]
        parse = map (map read . splitOn "\t") . lines
        permutations :: Integral d => d -> [a] -> [[a]]
        permutations n options = filter (\permutation -> genericLength permutation == n) $ replicateM (fromIntegral n) options
 
    describe "table" $ do
      let test title predicate = do
            describe title $ do
              it "formats the truth table of the predicate correctly" $ do
                hPutStrLn stderr $ unlines [title, "A\tB\t" ++ title, table predicate]
                (table predicate & parse) `shouldBe` [[a, b, predicate a b] | a <- [False, True], b <- [False, True]]
      test "and'" and'
      test "or'" or'
      test "nand'" nand'
      test "nor'" nor'
      test "xor'" xor'
      test "impl'" impl'
      test "equ'" equ'

    -- Problem 48
    describe "tablen" $ do
      let alphabets = map (\c -> [c]) ['A'..] :: [String]
          test title predicate n = do
            describe title $ do
              it "formats the truth table of the predicate correctly" $ do
                hPutStrLn stderr $ unlines [title, (intercalate "\t" $ genericTake n alphabets) ++ "\t" ++ title, tablen n predicate]
                (tablen n predicate & parse) `shouldBe` (map (\input -> input ++ [predicate input]) $ replicateM n [False, True])

      let predicate [a, b, c] = a `and'` (b `or'` c) `equ'` a `and'` b `or'` a `and'` c
      test "A `and'` (B `or'` C) `equ'` A `and'` B `or'` A `and'` C" predicate 3


  -- Problem 49
  describe "gray" $ do
    it "produces an n-bit Gray code" $ do
      gray 1 `shouldBe` ["0", "1"]
      gray 2 `shouldBe` ["00", "01", "11", "10"]
      gray 3 `shouldBe` ["000", "001", "011", "010", "110", "111", "101", "100"]


src/Lib.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
module Lib where

import System.Random

import Control.Arrow (second)
import Control.Monad (replicateM)

import qualified Data.List as List
import Data.Ord (comparing)
import Data.Map (Map, (!))
import qualified Data.Map as Map

import Debug.Trace

sniff input = trace (show input) input

-- | Problem 1: Find the last element of a list.
myLast :: [a] -> a
myLast [] = error "empty list"
myLast (x:[]) = x
myLast (_:xs) = myLast xs

-- | Problem 2: Find the last but one element of a list.
myButLast :: [a] -> a
myButLast [] = error "empty list"
myButLast (_:[]) = error "list with one element"
myButLast (x:_:[])  = x
myButLast (_:xs) = myButLast xs

-- | Problem 3: Find the K'th element of a list. The first element in the list is number 1.
elementAt :: Integral d => [a] -> d -> a
elementAt [] _ = error "empty list"
elementAt (x:_) 1 = x
elementAt (_:xs) index = elementAt xs (index - 1)

-- | Problem 4: Find the number of elements of a list.
myLength :: Integral d => [a] -> d
myLength [] = 0
myLength (_:xs) = 1 + myLength xs

-- | Problem 5: Reverse a list.
myReverse :: [a] -> [a]
myReverse [] = []
myReverse (x:xs) = myReverse xs ++ [x]

-- | Problem 6: Find out whether a list is a palindrome. A palindrome can be read forward or backward; e.g. (x a m a x).
isPalindrome :: Eq a => [a] -> Bool
isPalindrome xs = xs == reverse xs 

-- | Problem 7: Flatten a nested list structure.
-- Transform a list, possibly holding lists as elements into a `flat' list by replacing each list with its elements (recursively).
data NestedList a = Elem a | List [NestedList a]
flatten :: NestedList a -> [a]
flatten (Elem x) = [x]
flatten (List xs) = concat $ map flatten xs
  

-- | Problem 8: Eliminate consecutive duplicates of list elements.
compress :: Eq a => [a] -> [a]
compress [] = []
compress [x] = [x]
compress (x:x':xs)
  | x == x' = compress $ x:xs
  | otherwise = x:(compress $ x':xs)
-- compress = map head . group

-- | Problem 9: Pack consecutive duplicates of list elements into sublists. If a list contains repeated elements they should be placed in separate sublists.
pack :: Eq a => [a] -> [[a]]
pack [] = []
pack [x] = [[x]]
pack (x:x':xs)
  | x == head group = ([x] ++ group):groups
  | otherwise = [x]:group:groups
  where
    group:groups = pack (x':xs)

-- | Problem 10: Run-length encoding of a list. Use the result of problem P09 to implement the so-called run-length encoding data compression method. Consecutive duplicates of elements are encoded as lists (N E) where N is the number of duplicates of the element E.
encode :: Eq a => [a] -> [(Int, a)]
encode xs = map (\group -> (length group, head group)) $ pack xs

-- | Problem 11: Modified run-length encoding.
-- Modify the result of problem 10 in such a way that if an element has no duplicates it is simply copied into the result list. Only elements with duplicates are transferred as (N E) lists.
data ModifiedEncoding a
  = Single a
  | Multiple Int a
  deriving (Show, Eq)

modifiedEncoding :: (Int, a) -> ModifiedEncoding a
modifiedEncoding (1, x) = Single x
modifiedEncoding (n, x) = Multiple n x

encodeModified :: Eq a => [a] -> [ModifiedEncoding a]
encodeModified xs = map modifiedEncoding $ encode xs

-- | Problem 12: Decode a run-length encoded list.
-- Given a run-length code list generated as specified in problem 11. Construct its uncompressed version.
decodeModified :: [ModifiedEncoding a] -> [a]
decodeModified encoded = concat $ map decode encoded
  where
    decode (Single x) = [x]
    decode (Multiple n x) = replicate n x

-- | Problem 13: Run-length encoding of a list (direct solution).
-- Implement the so-called run-length encoding data compression method directly. I.e. don't explicitly create the sublists containing the duplicates, as in problem 9, but only count them. As in problem P11, simplify the result list by replacing the singleton lists (1 X) by X.
encodeDirect :: Eq a => [a] -> [ModifiedEncoding a]
encodeDirect = map modifiedEncoding . encode'
  where
    encode' = foldr f []
    f x [] = [(1, x)]
    f x ((n, y):ys)
      | x == y = (n + 1, x):ys
      | otherwise = (1, x):(n, y):ys 

-- | Problem 14: Duplicate the elements of a list.
dupli :: [a] -> [a]
dupli [] = []
dupli (x:xs) = x:x:dupli xs

-- | Problem 15: Replicate the elements of a list a given number of times.
repli :: [a] -> Int -> [a]
repli [] _ = []
repli (x:xs) n = (take n $ repeat x) ++ (repli xs n)

-- | Problem 16: Drop every N'th element from a lis.
dropEvery :: Integral d => [a] -> d -> [a]
dropEvery xs n = unenumerate $ filter (\(i, _) -> i `mod` n /= (n - 1)) $ enumerate xs
  where
    enumerate = zip [0..]
    unenumerate = map snd

-- | Problem 17: Split a list into two parts; the length of the first part is given.

-- Do not use any predefined predicates.
split :: [a] -> Int -> ([a], [a])
split xs n = helper ([], xs) n
  where
    helper (xs, ys) 0 = (xs, ys)
    helper (xs, []) _ = (xs, [])
    helper (xs, y:ys) n = helper (xs ++ [y], ys) (n - 1)

-- | Problem 18: Extract a slice from a list.
-- Given two indices, i and k, the slice is the list containing the elements between the i'th and k'th element of the original list (both limits included). Start counting the elements with 1.
slice :: [a] -> Int -> Int -> [a]
slice xs lo hi = drop (lo - 1) $ take hi xs

-- | Problem 19: Rotate a list N places to the left.
-- Hint: Use the predefined functions length and (++).
rotate :: [a] -> Int -> [a]
rotate xs n = drop n' xs ++ take n' xs
  where
    n' = (length xs + n) `mod` (length xs)

-- | Problem 20: Remove the K'th element from a list.
removeAt :: Int -> [a] -> (a, [a])
removeAt k xs = (xs !! (k - 1), take (k - 1) xs ++ drop k xs)

-- | Problem 21: Insert an element at a given position into a list.
insertAt :: a -> [a] -> Int -> [a]
insertAt x xs index = take (index - 1) xs ++ [x] ++ drop (index - 1) xs

-- | Problem 22: Create a list containing all integers within a given range.
range :: Enum a => a -> a -> [a]
range lo hi = [lo..hi]

-- | Problem 23: Extract a given number of randomly selected elements from a list.
rnd_select :: [a] -> Int -> IO [a]
rnd_select xs n
  | n < 0 = error "n has to be > 0"
  | otherwise = do
      generator <- getStdGen
      return $ take n [xs !! index | index <- randomRs (0, length xs - 1)  generator]

-- | Problem 24: Lotto: Draw N different random numbers from the set 1..M.
diff_select :: Int -> Int -> IO [Int]
diff_select n m = rnd_select [1..m] n

-- | Problem 25: Generate a random permutation of the elements of a list.
rnd_permu :: [a] -> IO [a]
rnd_permu xs = rnd_select xs (length xs)

-- | Problem 26: Generate the combinations of K distinct objects chosen from the N elements of a list
-- In how many ways can a committee of 3 be chosen from a group of 12 people? We all know that there are C(12,3) = 220 possibilities (C(N,K) denotes the well-known binomial coefficients). For pure mathematicians, this result may be great. But we want to really generate all the possibilities in a list.
combinations :: Int -> [a] -> [[a]]
combinations k xs = backtrack xs [] (length xs) k 0
  where
    backtrack :: [a] -> [a] -> Int -> Int -> Int -> [[a]]
    backtrack xs current n k start
      | length current == k = [current]
      | otherwise = concat $ [backtrack xs (current ++ [xs !! i]) n k (i + 1) | i <- [start..(n - 1)]]

-- | Problem 27: Group the elements of a set into disjoint subsets.
group :: [Int] -> [a] -> [[[a]]]
group [] _ = [[]]
group (n:ns) xs = [g:gs | (g, rs) <- combination n xs, gs <- group ns rs]
  where
    combination 0 xs = [([], xs)]
    combination _ [] = []
    combination n (x:xs) = [(x:ys, zs) | (ys, zs) <- combination (n - 1) xs] ++ [(ys, x:zs) | (ys, zs) <- combination n xs]

-- | Problem 28 (a): Sorting a list of lists according to length of sublists
lsort :: [[a]] -> [[a]]
lsort = List.sortOn length

-- | Problem 28 (b): Sorting a list of lists according to length of frequency sublists
lfsort :: [[a]] -> [[a]]
lfsort xs = List.sortOn (\x -> frequencies ! (length x)) xs
  where
    frequencies :: Map Int Int
    frequencies = foldr (\x -> Map.insertWith (+) (length x) 1) Map.empty xs

-- | Problem 31: Determine whether a given integer number is prime.
isPrime :: Integral d => d -> Bool
isPrime n = List.find (>= n) primes == Just n

primes :: Integral d => [d]
primes = sieve_of_eratosthenes [2..]

sieve_of_eratosthenes :: Integral d => [d] -> [d]
sieve_of_eratosthenes [] = undefined
sieve_of_eratosthenes (x:xs) = x:sieve_of_eratosthenes [i | i <- xs, i `mod` x > 0]

-- | Problem 32: Determine the greatest common divisor of two positive integer numbers. Use Euclid's algorithm.
myGCD :: Integral d => d -> d -> d
myGCD a b
  | b == 0 = abs a
  | otherwise = myGCD b (a `mod` b)

-- | Problem 33: Determine whether two positive integer numbers are coprime. Two numbers are coprime if their greatest common divisor equals 1.
coprime :: Integral d => d -> d -> Bool
coprime a b = gcd a b == 1

-- | Problem 34: Calculate Euler's totient function phi(m).
-- Euler's so-called totient function phi(m) is defined as the number of positive integers r (1 <= r < m) that are coprime to m.
totient :: Integral d => d -> d
totient m = List.genericLength $ [r | r <- [1..(m - 1)], coprime r m]

-- | Problem 35: Determine the prime factors of a given positive integer. Construct a flat list containing the prime factors in ascending order.
primeFactors :: Integral d => d -> [d]
primeFactors n = concat [List.genericReplicate (multiplicity n p) p | p <- takeWhile (<= n) primes, n `mod` p == 0]

sqrt' :: Integral d => d -> d
sqrt' n = floor $ sqrt $ fromIntegral n

multiplicity :: Integral d => d -> d -> d
multiplicity dividend divisor = case dividend `divMod` divisor of
  (quotient, 0) -> 1 + multiplicity quotient divisor
  _ -> 0 

-- | Problem 36: Determine the prime factors of a given positive integer.
prime_factors_mult :: Integral d => d -> [(d, d)]
prime_factors_mult n = [(p, multiplicity n p) | p <- takeWhile (<= n) primes, n `mod` p == 0]

-- | Problem 37: Calculate Euler's totient function phi(m) (improved)
-- See problem 34 for the definition of Euler's totient function. If the list of the prime factors of a number m is known in the form of problem 36 then the function phi(m) can be efficiently calculated as follows: Let ((p1 m1) (p2 m2) (p3 m3) ...) be the list of prime factors (and their multiplicities) of a given number m. Then phi(m) can be calculated with the following formula:
-- 
-- phi(m) = (p1 - 1) * p1 ** (m1 - 1) * 
--          (p2 - 1) * p2 ** (m2 - 1) * 
--          (p3 - 1) * p3 ** (m3 - 1) * ...
-- Note that a ** b stands for the b'th power of a.
totient' :: Integral d => d -> d
totient' m = product $ [(p - 1) * p ^ (k - 1) | (p, k) <- prime_factors_mult m]

-- | Problem 38: Compare the two methods of calculating Euler's totient function.
-- Use the solutions of problems 34 and 37 to compare the algorithms. Take the number of reductions as a measure for efficiency. Try to calculate phi(10090) as an example.
-- (no solution required)
-- NOTE: gcd(a, b) has a time complexity of O(log(min(a, b)))
-- So coprime is also of O(log(min(a, b)))
-- totient from (34) is then of O(m^2 * log(m))
-- totient' from (37) is of O(m^2)

-- | Problem 39: A list of prime numbers.
-- Given a range of integers by its lower and upper limit, construct a list of all prime numbers in that range.
primesR :: Integral d => d -> d -> [d]
primesR lo hi = takeWhile (<= hi) $ dropWhile (<= lo) $ primes

-- | Problem 40: Goldbach's conjecture.
-- Goldbach's conjecture says that every positive even number greater than 2 is the sum of two prime numbers. Example: 28 = 5 + 23. It is one of the most famous facts in number theory that has not been proved to be correct in the general case. It has been numerically confirmed up to very large numbers (much larger than we can go with our Prolog system). Write a predicate to find the two prime numbers that sum up to a given even integer.
goldbach :: Integral d => d -> (d, d)
goldbach n | n <= 2 || odd n = undefined
goldbach n = head [(x, y) | x <- possibilities, y <- filter (>= x) possibilities, x + y == n]
  where
    possibilities = takeWhile (<= n) primes

-- | Problem 41: Given a range of integers by its lower and upper limit, print a list of all even numbers and their Goldbach composition.
-- In most cases, if an even number is written as the sum of two prime numbers, one of them is very small. Very rarely, the primes are both bigger than say 50. Try to find out how many such cases there are in the range 2..3000.
goldbachList :: Integral d => d -> d -> [(d, d)]
goldbachList lo hi = map goldbach $ takeWhile (<= hi) $ dropWhile (<= lo) evens
  where
    evens :: Integral d => [d]
    evens = [2, 4..]

goldbachList' :: Integral d => d -> d -> d -> [(d, d)]
goldbachList' lo hi limit = [(a, b) | (a, b) <- goldbachList lo hi, a > limit && b > limit]

-- | Problem 46: Define predicates and/2, or/2, nand/2, nor/2, xor/2, impl/2 and equ/2 (for logical equivalence) which succeed or fail according to the result of their respective operations; e.g. and(A,B) will succeed, if and only if both A and B succeed.
-- A logical expression in two variables can then be written as in the following example: and(or(A,B),nand(A,B)).
-- Now, write a predicate table/3 which prints the truth table of a given logical expression in two variables.
type Predicate = Bool -> Bool -> Bool

infixr 3 `and'`
and' :: Predicate
True `and'` True = True
_ `and'` _ = False

infixr 2 `or'`
or' :: Predicate
False `or'` False = False
_ `or'` _ = True

not' :: Bool -> Bool
not' True = False
not' False = True

infixr 3 `nand'`
nand' :: Predicate
a `nand'` b = not' $ a `and'` b

infixr 2 `nor'`
nor' :: Predicate
a `nor'` b = not' $ a `or'` b

infixr 1 `xor'`
xor' :: Predicate
True `xor'` False = True
False `xor'` True = True
_ `xor'` _ = False

-- | a implies b
impl' :: Predicate
a `impl'` b = (not' a) `or'` b

infixr 0 `equ'`
equ' :: Predicate
True `equ'` True = True
False `equ'` False = True
_ `equ'` _ = False

-- | Truth Table
table :: (Predicate) -> String
table predicate = unlines [List.intercalate "\t" $ map show $ [a, b, predicate a b] | a <- [False, True], b <- [False, True]]

-- | Problem 47: Truth tables for logical expressions (2).
-- Continue problem P46 by defining and/2, or/2, etc as being operators. This allows to write the logical expression in the more natural way, as in the example: A and (A or not B). Define operator precedence as usual; i.e. as in Java.
-- (Done with Problem 46)

-- | Problem 48: Truth tables for logical expressions (3).
-- Generalize problem P47 in such a way that the logical expression may contain any number of logical variables. Define table/2 in a way that table(List,Expr) prints the truth table for the expression Expr, which contains the logical variables enumerated in List.
tablen :: Integral d => d -> ([Bool] -> Bool) -> String
tablen n predicate = unlines $ [List.intercalate "\t" $ map show $ input ++ [predicate input] | input <- replicateM (fromIntegral n) [False, True]]

-- | Problem 49: Gray Code
gray :: Integral d => d -> [String]
gray n | n < 1 = undefined
gray 1 = ["0", "1"]
gray n = ['0':code | code <- previous] ++ ['1':code | code <- reverse previous]
  where
    previous = gray (n - 1)

-- | Problem 50: Huffman codes.
-- We suppose a set of symbols with their frequencies, given as a list of fr(S,F) terms. Example: [fr(a,45),fr(b,13),fr(c,12),fr(d,16),fr(e,9),fr(f,5)]. Our objective is to construct a list hc(S,C) terms, where C is the Huffman code word for the symbol S. In our example, the result could be Hs = [hc(a,'0'), hc(b,'101'), hc(c,'100'), hc(d,'111'), hc(e,'1101'), hc(f,'1100')] [hc(a,'01'),...etc.].
huffman :: Integral d => [(Char, d)] -> [(Char, String)]
huffman = shrink . map (\(c, p) -> (p, [(c ,"")])) . List.sortOn snd
  where
    shrink [(_, xs)] = List.sortOn fst xs
    shrink (x:x':xs) = shrink $ insertOn fst (add x x') xs
    add (p, xs) (p', xs') = (p + p', prefixing '0' xs ++ prefixing '1' xs')
    prefixing c = map (second (c:))
    insertOn f = List.insertBy (comparing f)
